<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>coR-ge by Chris1221</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-light.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>coR-ge</h1>
        <p>Software for the Examination of Multiple Correction Methodologies in Accurate Genomic Environments</p>

        <p class="view"><a href="https://github.com/Chris1221/sFDR">View the Project on GitHub <small>Chris1221/sFDR</small></a></p>


        <ul>
          <li><a href="https://github.com/Chris1221/sFDR/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/Chris1221/sFDR/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/Chris1221/sFDR">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h2>
<a id="quick-start-guide" class="anchor" href="#quick-start-guide" aria-hidden="true"><span class="octicon octicon-link"></span></a>Quick-start Guide.</h2>

<p>Welcome to the landing page for <strong>coR-ge</strong> (<strong>cor</strong>ection of <strong>ge</strong>nomes in <strong>R</strong>). This software is in active development, and pull requests are welcome on any of the branches. Please find below instructions for use cases. For more complex use, please contact the maintainer by <a href="https://github.com/Chris1221/sFDR/issues/new">raising an issue</a> on the project repository.</p>

<p>This project was presented at the Compute Canada's High Powered Computer Symposium in June 2015. The <a href="https://github.com/Chris1221">abstract</a> and <a href="https://github.com/Chris1221">poster PDF</a> are also found in the repository. </p>

<h3>
<a id="program-structure" class="anchor" href="#program-structure" aria-hidden="true"><span class="octicon octicon-link"></span></a>Program Structure</h3>

<p>Heuristically, <strong>coR-ge</strong> is structured as follows.  </p>

<p><img src="https://raw.githubusercontent.com/Chris1221/sFDR/gh-pages/assets/images/logic_flow.png" alt="Logic Flow"></p>

<p>In reality, these three steps are divided into different files in the repository.  The quick start guide will nonetheless follow this logic flow in order to simplify matters.   </p>

<h3>
<a id="requirements" class="anchor" href="#requirements" aria-hidden="true"><span class="octicon octicon-link"></span></a>Requirements</h3>

<p><strong>coR-ge</strong> requires the following: </p>

<ul>
<li>This program requires a cluster system set up on either <code>SGE</code> or <code>PBS</code> architecture, running <code>qsub</code>.  It can be easily modified to use <code>bsub</code> as well with minimal effort.</li>
<li>Any base unix system to run on.  The program has been tested on <code>CentOS</code> and <code>OSX</code>.  Dependency issues should be reported as an <a href="https://github.com/Chris1221/sFDR/issues/new">issue</a> on the main repository. </li>
<li>
<code>R</code> version &gt; 3.0.0 able to run <code>dplyr</code> and <code>data.frame</code>
</li>
<li>
<code>Rscript</code> in <code>$PATH</code> or located at <code>/usr/bin/Rscript</code>
</li>
<li>
<code>Python</code> v. 2.7.*</li>
<li>A temporary storage drive with n*560G free space, where n is the number of runs you would like to perform simultaneously. </li>
<li>156G RAM recommended. Be aware that improper allocation of jobs can easily cause swapping of nodes on the cluster.  We have included a delayed submission script to ameliorate this problem. </li>
<li>Latest build of 1000 Genomes in IMPUTE2 format (<code>*.legend</code>, <code>*.haps</code>, <code>*.sample</code>) in any directory that has sufficient storage room.  Name this directory; it will hereafter be named <code>$REF</code>
</li>
</ul>

<h3>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h3>

<p>Clone the repository to your working directory (hereafter known as <code>$WD</code>):</p>

<pre><code>$ cd $WD
$ git clone https://github.com/Chris1221/sFDR
$ git checkout master
$ cd sFDR/
</code></pre>

<p>Ensure that you are on the master branch for now.  </p>

<p>Several files will have to be edited, and these will be addressed in turn. </p>

<h3>
<a id="disease-model" class="anchor" href="#disease-model" aria-hidden="true"><span class="octicon octicon-link"></span></a>Disease Model</h3>

<p>The file <code>rand.R</code> contains the information for the disease model. We will edit a few things here to specify individual parameters for the run.  At the top of the script, you must edit the following parameters.  Defaults are displayed, and must be edited:</p>

<pre><code>heritability &lt;- 0.45
noise_var &lt;- 0.55
sim_file_path &lt;- getwd()
wd &lt;- getwd()
snp_summary &lt;- "snp_summary.out"
</code></pre>

<p>To filter based on minor allele frequency, you must generate a file named <code>snp_summary.out</code> (or whatever you set the file name as in <code>rand.R</code>) using SNPTEST2. See <a href="https://mathgen.stats.ox.ac.uk/genetics_software/snptest/old/snptest.html#Data_Summaries_">here</a> for an example of this.</p>

<p>In this script, you can manually or randomly generate causal loci (~ line 160).  If you would like to manually assign variants, choose your SNPs and assign them a column named "causal" equal to one, and pass only causal SNPs to the <code>snps</code> data frame.  <code>snps</code> must contain four columns, an example is given below: </p>

<pre><code>rsid chromosomes bp all_maf
rs_1 1           1  0.30
rs_2 1           2  0.35
rs_3 1           3  0.05
</code></pre>

<p>If the selection of the SNPs is based on some parameter, simply subset the <code>summary</code> object to select only causal loci.  Ex:</p>

<pre><code>library(dplyr)
summary %&gt;%
    filter(all_maf &gt; 0.05) %&gt;% 
        filter(all_maf &lt; 0.5) %&gt;% 
            sample_n(50) %&gt;% 
                select(rsid, chromosomes, bp, all_maf) -&gt; snps
</code></pre>

<h3>
<a id="resampling-and-model-propagation" class="anchor" href="#resampling-and-model-propagation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Resampling and Model Propagation</h3>

<p>In <code>i_array.sh</code>, <code>j_array.sh</code>, and <code>k_array.sh</code>, set the <code>$DIR</code> variable with the directory where you are storing the reference genotypes. Set the <code>$OUT</code> variable for where you would like to house the output. Edit the directives to fit your cluster set-up.</p>

<p>Start the generation of resampled genotypes with <code>hapgen2</code>.  For 30 trials (approximately 1.7 TB of free space required):</p>

<pre><code>for i in {1..3}
do
  for j in {1..10}
  do
    bash i_array $i $j
  done
done
</code></pre>

<p>or
    bash sub_script.sh $i_limit $j_limit
for ease of use. </p>

<p>This will begin the submission of many small jobs.  If any jobs fail on your system, please post the full log file in an issue.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/Chris1221">Chris1221</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
              <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-63747419-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
